File: Python/knowledge-base/knowledge_base/Ingestion/FileIngester.py
from pathlib import Path
from typing import Dict, Any, Optional


class FileIngester:
    """Reads supported file types and returns a dict with document fields."""

    SUPPORTED_EXTENSIONS = {".txt", ".md", ".pdf"}

    def ingest_file(self, file_path: Path) -> Optional[Dict[str, Any]]:
        """
        Read a file and return document fields.

        Args:
            file_path: Path to the file to ingest.

        Returns:
            Dict with keys: title, source_path, source_type, raw_content,
            metadata. Returns None on failure.
        """
        file_path = Path(file_path)

        if not file_path.exists():
            print(f"File not found: {file_path}")
            return None

        suffix = file_path.suffix.lower()

        if suffix not in self.SUPPORTED_EXTENSIONS:
            print(f"Unsupported file type: {suffix}")
            return None

        if suffix in {".txt", ".md"}:
            return self._ingest_text_file(file_path)
        elif suffix == ".pdf":
            return self._ingest_pdf_file(file_path)

        return None

    def _ingest_text_file(self, file_path: Path) -> Optional[Dict[str, Any]]:
        """Read a plain text or markdown file."""
        try:
            raw_content = file_path.read_text(encoding="utf-8")
            source_type = "markdown" if file_path.suffix.lower() == ".md" \
                else "text"
            return {
                "title": file_path.stem,
                "source_path": str(file_path.resolve()),
                "source_type": source_type,
                "raw_content": raw_content,
                "metadata": {
                    "filename": file_path.name,
                    "size_bytes": file_path.stat().st_size
                }
            }
        except Exception as e:
            print(f"Error reading text file {file_path}: {e}")
            return None

    def _ingest_pdf_file(self, file_path: Path) -> Optional[Dict[str, Any]]:
        """Read a PDF file using pypdf (or PyPDF2 as fallback)."""
        try:
            try:
                from pypdf import PdfReader
            except ImportError:
                from PyPDF2 import PdfReader  # type: ignore

            reader = PdfReader(str(file_path))
            pages_text = []
            for page in reader.pages:
                text = page.extract_text()
                if text:
                    pages_text.append(text)

            raw_content = "\n".join(pages_text)
            return {
                "title": file_path.stem,
                "source_path": str(file_path.resolve()),
                "source_type": "pdf",
                "raw_content": raw_content,
                "metadata": {
                    "filename": file_path.name,
                    "size_bytes": file_path.stat().st_size,
                    "num_pages": len(reader.pages)
                }
            }
        except ImportError:
            print(
                "pypdf/PyPDF2 not installed. Install pypdf to support PDF files.")
            return None
        except Exception as e:
            print(f"Error reading PDF file {file_path}: {e}")
            return None

File: Rust/knowledge-base/src/ingestion/file_ingester.rs
use anyhow::{bail, Context, Result};
use sha2::{Digest, Sha256};
use std::path::Path;

/// The result of ingesting a file — raw document fields ready for DB insertion.
#[derive(Debug, Clone)]
pub struct IngestedDocument {
    pub title: String,
    pub source_path: String,
    pub source_type: String,
    pub raw_content: String,
    pub metadata: Option<serde_json::Value>,
}

/// Reads supported file types (.txt, .md) and returns document fields.
#[derive(Debug, Default)]
pub struct FileIngester;

impl FileIngester {
    pub fn new() -> Self {
        Self
    }

    /// Read a file and return an `IngestedDocument`.
    ///
    /// Supported extensions: `.txt`, `.md`
    /// Returns `Err` for unsupported file types or I/O failures.
    pub fn ingest_file(path: &Path) -> Result<IngestedDocument> {
        if !path.exists() {
            bail!("File not found: {}", path.display());
        }

        let extension = path
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_lowercase();

        match extension.as_str() {
            "txt" | "md" => Self::ingest_text_file(path),
            "pdf" => Self::ingest_pdf_file(path),
            other => bail!("Unsupported file type: .{}", other),
        }
    }

    fn ingest_text_file(path: &Path) -> Result<IngestedDocument> {
        let raw_content = std::fs::read_to_string(path)
            .with_context(|| format!("Failed to read file: {}", path.display()))?;

        let source_type = if path
            .extension()
            .and_then(|e| e.to_str())
            .map(|e| e.eq_ignore_ascii_case("md"))
            .unwrap_or(false)
        {
            "markdown".to_string()
        } else {
            "text".to_string()
        };

        let title = path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("untitled")
            .to_string();

        let source_path = path
            .canonicalize()
            .unwrap_or_else(|_| path.to_path_buf())
            .to_string_lossy()
            .to_string();

        let size_bytes = path.metadata().map(|m| m.len()).unwrap_or(0);
        let filename = path
            .file_name()
            .and_then(|f| f.to_str())
            .unwrap_or("")
            .to_string();

        let metadata = serde_json::json!({
            "filename": filename,
            "size_bytes": size_bytes,
        });

        Ok(IngestedDocument {
            title,
            source_path,
            source_type,
            raw_content,
            metadata: Some(metadata),
        })
    }

    fn ingest_pdf_file(path: &Path) -> Result<IngestedDocument> {
        let raw_content = pdf_extract::extract_text(path)
            .with_context(|| format!("Failed to extract text from PDF: {}", path.display()))?;

        if raw_content.trim().is_empty() {
            bail!(
                "PDF contains no extractable text (scanned/image-only?): {}",
                path.display()
            );
        }

        let title = path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("untitled")
            .to_string();

        let source_path = path
            .canonicalize()
            .unwrap_or_else(|_| path.to_path_buf())
            .to_string_lossy()
            .to_string();

        let size_bytes = path.metadata().map(|m| m.len()).unwrap_or(0);
        let filename = path
            .file_name()
            .and_then(|f| f.to_str())
            .unwrap_or("")
            .to_string();

        let metadata = serde_json::json!({
            "filename": filename,
            "size_bytes": size_bytes,
        });

        Ok(IngestedDocument {
            title,
            source_path,
            source_type: "pdf".to_string(),
            raw_content,
            metadata: Some(metadata),
        })
    }

    /// Compute the SHA-256 hex digest of a string.
    pub fn compute_sha256(content: &str) -> String {
        let mut hasher = Sha256::new();
        hasher.update(content.as_bytes());
        hex::encode(hasher.finalize())
    }
}
